### PR TITLE
Demos/appnp

### PR DESCRIPTION
**Pull request by kieranricardo**
I've added in an APPNP demo for the reddit dataset from the GraphSAGE paper and included APPNP in the twitter demo.

One issue is:

- APPNP doesn't naturally allow for node embeddings. To keep the workflow of the twitter demo consistent I obtained node embeddings by propagating the MLP features (as opposed to the MLP predictions) but the resulting embeddings aren't well separated by TSNE=

#### Reviewer Checklist
- [x] Twitter demo notebook runs and clearly demonstrates how to use APPNP
~- [ ] Feature propagation in twitter notebook is correct~
- [x] Reddit demo notebook runs and clearly demonstrates how to use APPNP

### CODE REVIEW

**Comment by kieranricardo** [comment]
hi @habiba-h, I've added the new SG class in and the memory use is way down. I've also removed appnp from hateful twitters because I don't think it added much and made it more confusing. I think its ready for review again :)

File: demos/node-classification/ppnp/appnp-reddit.ipynb [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+    ">20 days for training and the remaining days for testing (with 30% used for validation). For features,\n",
+    ">we use off-the-shelf 300-dimensional GloVe CommonCrawl word vectors [3]; for each post, we\n",
+    ">concatenated (i) the average embedding of the post title, (ii) the average embedding of all the post’s\n",
+    ">comments (iii) the post’s score, and (iv) the number of comments made on the post.\n",
+    "\n",
+    "\n",
+    "We demonstrate the advantage of using graph features and the scalability of the APPNP algorithm for node classification on the reddit dataset.  We first train a MLP on the node features and then propagate this model using APPNP.  Training is only done on the node features, this approach allows model training to be completed in under a 1 minute on an 8th gen quad-core i7.\n",
+    "\n",
+    "\n",
+    "**References**\n",
```
Comment (habiba-h) [inline demos/node-classification/ppnp/appnp-reddit.ipynb]: It would be nice, if you add hyperlinks to the references.

File: demos/node-classification/ppnp/appnp-reddit.ipynb [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+    "target_encoding = preprocessing.OneHotEncoder(sparse=False, categories=\"auto\")\n",
+    "targets = target_encoding.fit_transform(labels.reshape(-1, 1))\n",
+    "targets = pd.DataFrame(targets)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We then split the data into train/val/test based on the labels stored in the dataset. This is the same train/val/test split used in the graphsage paper. Then we fit a standard scaler on only the training data and use it to standardize all of the data."
```
Comment (habiba-h) [inline demos/node-classification/ppnp/appnp-reddit.ipynb]: Maybe add the link to the GraphSAGE paper. 
Also, is this the same split as in the description above? Just checking. 
Talking about this from the description at the top 

> We use the first 20 days for training and the remaining days for testing (with 30% used for validation).

File: demos/node-classification/ppnp/appnp-reddit.ipynb [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+    "print(\"{} training nodes.\".format(len(train_data)))\n",
+    "print(\"{} validation nodes.\".format(len(val_data)))\n",
+    "print(\"{} testing nodes.\".format(len(test_data)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Creating an adjacnecy matrix and StellarGraph generator\n",
```
Comment (habiba-h) [inline demos/node-classification/ppnp/appnp-reddit.ipynb]: A StellarGraph graph not 

> an adjacnecy matrix

File: demos/node-classification/ppnp/appnp-reddit.ipynb [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+    "print(\"\\nTest Set Metrics:\")\n",
+    "for name, val in zip(propagated_model.metrics_names, test_metrics):\n",
+    "    print(\"\\t{}: {:0.4f}\".format(name, val))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Propagating the MLP with APPNP increases the test set accuracy by ~15% without any further training. As we are performing single-label multiclass classification the accuracy is equivalent to the micro F1 metric. This micro F1 is comparable to that attained in the GraphSAGE paper [2]. GraphSAGE with LSTM aggregation attains a best supervised F1 of 0.95 in [2], however APPNP only required ~3 minutes of training on an 8th gen i7 compared to the hours required for GraphSAGE while still attaining a similar F1."
```
Comment (habiba-h) [inline demos/node-classification/ppnp/appnp-reddit.ipynb]: I think this is the punchline, maybe make it bold!

File: demos/node-classification/ppnp/appnp-reddit.ipynb [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+     "output_type": "stream",
+     "text": []
+    }
+   ],
+   "source": [
+    "in_layer = layers.Input(shape=(train_data.shape[-1]))\n",
+    "\n",
+    "layer = layers.Dense(512, activation=\"relu\", kernel_regularizer=\"l2\")(in_layer)\n",
+    "layer = layers.Dropout(0.5)(layer)\n",
+    "layer = layers.Dense(512, activation=\"relu\", kernel_regularizer=\"l2\")(in_layer)\n",
```
Comment (habiba-h) [inline demos/node-classification/ppnp/appnp-reddit.ipynb]: Just checking, the input to this `Dense` layer should be `layer` output from the previous layer or the `in_layer` is the correct one?

File: demos/node-classification/ppnp/appnp-reddit.ipynb (hunk 2 above)
Comment (kieranricardo) [reply demos/node-classification/ppnp/appnp-reddit.ipynb:161]: Good idea re the link :+1:

Yeah I'm pretty sure its the same split. I got the graph off of snap and the nodes were labelled as train, test, and val and I use those labels to split the data. So i'm guessing they're the same splits used in the paper

File: demos/node-classification/ppnp/appnp-reddit.ipynb (hunk 5 above)
Comment (kieranricardo) [reply demos/node-classification/ppnp/appnp-reddit.ipynb:266]: woops, nice catch! I'll fix that 👍

File: demos/node-classification/ppnp/appnp-reddit.ipynb (hunk 4 above)
Comment (kieranricardo) [reply demos/node-classification/ppnp/appnp-reddit.ipynb:390]: do you mean **bold** or bold? I can't tell 😆

Comment (habiba-h) [reply demos/node-classification/ppnp/appnp-reddit.ipynb:417]: > do you mean **bold** or bold? I can't tell 😆

**bold** 😀

**Comment by huonw** [comment]
Per our discussion at sprint planning, if we land this here without running it on CI, we should have a follow-up issue for getting it running (some approaches might be: just taking the download cost; using a smaller/reduced version of the datasets).

Alternatively, if we do https://github.com/stellargraph/stellargraph/issues/1221, we can land this in our "use cases" repo.
